### PR TITLE
Add Partitions Metadata Table

### PR DESCRIPTION
**Pull request by syun64**
TODO: Waiting on https://github.com/apache/iceberg-python/pull/599 to merge to introduce TimeTravel - Done

### CODE REVIEW

File: pyiceberg/table/__init__.py [hunk 1]
```diff
@@ -3410,6 +3411,94 @@ def _readable_metrics_struct(bound_type: PrimitiveType) -> pa.StructType:
             schema=entries_schema,
         )
 
+    def partitions(self) -> "pa.Table":
```
Comment (syun64) [inline pyiceberg/table/__init__.py:3414]: TODO: Waiting on https://github.com/apache/iceberg-python/pull/599 to merge to introduce TimeTravel

File: pyiceberg/table/__init__.py [hunk 2]
```diff
@@ -3410,6 +3411,94 @@ def _readable_metrics_struct(bound_type: PrimitiveType) -> pa.StructType:
...
+            pa.field('equality_delete_record_count', pa.int64(), nullable=False),
+            pa.field('equality_delete_file_count', pa.int32(), nullable=False),
+            pa.field('last_updated_at', pa.timestamp(unit='ms'), nullable=True),
+            pa.field('last_updated_snapshot_id', pa.int64(), nullable=True),
+        ])
+
+        partition_record = self.tbl.metadata.specs_struct()
+        has_partitions = len(partition_record.fields) > 0
+
+        if has_partitions:
```
Comment (syun64) [inline pyiceberg/table/__init__.py:3434]: Similar to how we did with the entries table, should we opt-in to keep consistent metadata table schema for both unpartitioned and partitioned tables?

**Review by Fokko — APPROVED** [review]
LGTM. Thanks for working on this @syun64

File: pyiceberg/table/__init__.py (hunk 2 above)
Comment (Fokko) [reply pyiceberg/table/__init__.py:3446]: I don't think that's needed. The schema will change anyway based on the partition spec.

I've checked and we have this behavior in Spark as well:

image

Comment (syun64) [reply pyiceberg/table/__init__.py:3446]: Sounds good @Fokko . That makes sense. Just wanted to double check if we wanted to carry this logic over.

**Review by HonahX — APPROVED** [review]
LGTM!

**Comment by HonahX** [comment]
Merged! Thanks @syun64 for working on this and thanks @Fokko for reviewing!

**Comment by jlhester** [comment]
Thanks @syun64 for implementing this! It would be pretty useful for some work I'm doing. When do you think this will be released? I see it's on the [near term roadmap](https://github.com/apache/iceberg-python/issues/736) but I don't have a sense of when 0.7.0 will be available.

**Comment by syun64** [comment]
> Thanks @syun64 for implementing this! It would be pretty useful for some work I'm doing. When do you think this will be released? I see it's on the [near term roadmap](https://github.com/apache/iceberg-python/issues/736) but I don't have a sense of when 0.7.0 will be available.

Sorry for the late response. We are actively working on finalizing the keystone features for 0.7.0 release - so I would say very soon 🤞

**Pull request merged**